### ISSUE TITLE
[ASK] Early stopping in grid search?

### ISSUE THREAD

**Issue by mmosc** [RUCAIBox/RecBole#1515]
Hi and thank you for this great library!

I was trying to perform a hyperparameter optimization using Hyperopt by simply running your `run_hyper.py` script. My config file looks like this:
```
dataset: ml-100k
model: BPR
data_path: ./dataset
valid_metric: 'NDCG@10'
```
and my params file like this:
```
learning_rate choice [1.e-1, 1.e-2, 1.e-3, 1.e-4]
embedding_size choice [64,96,128]
train_batch_size choice [512,1024,2048]
```

I am running with the `exhaustive` method, i.e., I am performing a grid search. What I noticed is that after 13 out of 36 iterations, I get the message 
```
29 Oct 18:20    INFO  Early stop triggered. Stopping iterations as condition is reach.
```
The same thing seems to be happening for some of your benchmark models, see [here](https://github.com/RUCAIBox/RecBole/blob/b616af6b6d4a8eaf6221477cc3d482048cc4f53e/benchmark/hyper/sequential/SINE-ml-1m_seq.md) for instance.

However, in my understanding early stopping should be applied during training, to avoid overfitting, and not during hyperparameter optimization, and definitely not in a grid search, since we are moving in a N-dimensional (with N=3 in this case) grid, and there is no ordering of the hyperparameters to establish whether we are moving away from the minimum or not. Even if there was a way, we are scanning over the grid, so the order with which we pick the points would affect which result would be considered as the best. Alternatively, early stopping would make sense to me if we are only looking at the learning rate as hyperparameter, for instance.

To sum up, could you explain why we are applying early stopping in a grid search with more than one hyperparameter?
# [EDIT]
In fact, if I change the early stop parameter from 10 to 9999 to ensure that the grid search keeps running, the best model is another one.
early stop 10:
```
best params:  {'embedding_size': 128, 'learning_rate': 0.0001, 'train_batch_size': 1024}
```
ealry stop 9999:
```
best params:  {'embedding_size': 128, 'learning_rate': 0.001, 'train_batch_size': 1024}
```

Best, 
Marta

**Comment by AoiDragon** [comment]
Hello @mmosc ,

Computing every combination of hyperparameters could be very time-cosuming, so we add `early stop` to let users decide whether consider all possibility or not. As you mentioned, users can set a large `early stop` to test on every combination.

**Comment by mmosc** [comment]
Okay, thank you for your reply.

As a follow up question: is early stopping used during training, then? In the standard way, to avoid overfitting.

**Comment by AoiDragon** [comment]
Yes, you are right.

**Comment by mmosc** [comment]
Great, thank you!! :blush:
